File: src/main/java/com/kolich/blog/git/BlogRepository.java
package com.kolich.blog.git;

import com.kolich.blog.BlogConfigurationFactory;
import com.kolich.curacao.annotations.Component;
import com.kolich.curacao.handlers.components.CuracaoComponent;
import org.apache.commons.io.FileUtils;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.lib.Repository;
import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
import org.slf4j.Logger;

import javax.servlet.ServletContext;
import java.io.File;

import static org.slf4j.LoggerFactory.getLogger;

@Component
public final class BlogRepository implements CuracaoComponent {

    private static final Logger logger__ = getLogger(BlogRepository.class);

    private static final Boolean isDevMode__ =
        BlogConfigurationFactory.isDevMode();
    private static final String blogRepoCloneUrl__ =
        BlogConfigurationFactory.getBlogRepoCloneUrl();
    private static final String clonePath__ =
        BlogConfigurationFactory.getClonePath();
    private static final Boolean shouldCloneFromScratchOnStartup__ =
        BlogConfigurationFactory.shouldCloneFromScratchOnStartup();

    private static final String userDir__ = System.getProperty("user.dir");

    private File clonePath_;
    private Repository repository_;

    @Override
    public void initialize(final ServletContext context) throws Exception {
        if(isDevMode__) {
            // If in "development mode", the clone path is the local copy of
            // the repo on disk.
            clonePath_ = new File(userDir__);
        } else {
            // Otherwise, it could be somewhere else either in an absolute
            // location (/...) or somewhere under the Servlet container when
            // relative.
            clonePath_ = (clonePath__.startsWith(File.separator)) ?
                // If the specified clone path in the application configuration
                // is "absolute", then use that path raw.
                new File(clonePath__) :
                // Otherwise, the clone path is relative to the container Servlet
                // context of this application.
                new File(context.getRealPath(File.separator + clonePath__));
        }
        logger__.info("Using repository clone path: " +
            clonePath_.getCanonicalFile());
        // If were not in dev mode, and the clone path doesn't exist or we need
        // to force clone from scratch, do that now...
        final boolean clone = (!clonePath_.exists() || shouldCloneFromScratchOnStartup__);
        if(!isDevMode__ && clone) {
            logger__.info("Clone path does not exist, or we were asked " +
                "to re-clone fresh on startup. So, cloning...");
            logger__.info("Using Git repo URL: " + blogRepoCloneUrl__);
            // Recursively delete the existing clone of the repo,
            // if it exists.
            FileUtils.deleteDirectory(clonePath_);
            // Clone the repository to disk.
            Git.cloneRepository()
                .setURI(blogRepoCloneUrl__)
                .setDirectory(clonePath_)
                .setBranch("master")
                .call();
        }
        // Construct a new pointer to the repository on disk.
        repository_ = new FileRepositoryBuilder()
            .setWorkTree(clonePath_)
            .readEnvironment()
            .build();
        if(!isDevMode__) {
            // Pull a fresh copy of the repo from the remote.  Yes, we may have
            // just cloned it fresh, but this is what we do, bandwidth is cheap.
            new Git(repository_).pull().call();
        }
        logger__.info("Successfully initialized Git repository: " +
            repository_);
    }

    @Override
    public void destroy(final ServletContext context) throws Exception {
        repository_.close();
    }

    public File getClonePath() {
        return clonePath_;
    }

    public Repository getContentRepository() {
        return repository_;
    }

}

File: src/main/scala/com/kolich/blog/git/BlogRepository.scala
package com.kolich.blog.git

import com.typesafe.scalalogging.slf4j.Logging
import com.kolich.curacao.handlers.components.CuracaoComponent
import javax.servlet.ServletContext
import com.kolich.blog.ApplicationConfig
import scala.reflect.io.{File, Path}
import org.eclipse.jgit.lib.Repository
import org.eclipse.jgit.api.Git
import org.eclipse.jgit.storage.file.FileRepositoryBuilder
import com.kolich.curacao.annotations.Component

@Component
final class BlogRepository extends CuracaoComponent with Logging {

  private[this] val devMode: Boolean = ApplicationConfig.isDevMode

  private[this] val clonePath: String = ApplicationConfig.clonePath
  private[this] val repoUrl: String = ApplicationConfig.blogRepoUrl
  private[this] val cloneFromScratchOnStartup: Boolean = ApplicationConfig.cloneFromScratchOnStartup

  var repoDir: Path = _
  var repo: Repository = _
  var git: Git = _

  override def initialize(context: ServletContext) = {
    // Discover the right path to the local repository on disk.
    repoDir = Path(devMode match {
      case true => System.getProperty("user.dir")
      case _ => clonePath.startsWith(File.separator) match {
        case true => clonePath
        case _ => context.getRealPath(File.separator + clonePath)
      }
    })
    logger.info("Using repository clone path: " + repoDir.toCanonical)
    val clone: Boolean = (repoDir.notExists || cloneFromScratchOnStartup)
    if(!devMode && clone) {
      logger.info("Clone path does not exist, or we were asked to re-clone" +
        "fresh on startup. So, cloning from: " + repoUrl)
      repoDir.deleteRecursively()
      Git.cloneRepository().setURI(repoUrl).setDirectory(repoDir.jfile)
        .setBranch("master").call
    }
    repo = new FileRepositoryBuilder().setWorkTree(repoDir.jfile)
      .readEnvironment.build
    git = new Git(repo)
    if(!devMode) {
      git.pull.call
    }
    logger.info("Successfully initialized Git repository: " + repo)
  }

  override def destroy(context: ServletContext) = {
    repo.close
  }

}
